packages/firestore/src/local/local_documents_view.ts

```diff
@@ -218,7 +221,43 @@ export class LocalDocumentsView {
         );
       })
       .next(matchingMutationBatches => {
+        mutationBatches = matchingMutationBatches;
+        // It is possible that a PatchMutation can make a document a match to the query,
+        // but the version in `remoteDocumentCache` is not a match yet (waiting for server
+        // to ack). To handle this, we find all document keys affected by some`PatchMutation`s
+        // that are not in `result` yet, and back fill them via `remoteDocumentCache.getEntries`,
+        // otherwise those `PatchMutations` will be ignored because no base document can be found,
+        // and lead to missing result for the query.
+        let missingBaseDocEntriesForPatching = documentKeySet();
         for (const batch of matchingMutationBatches) {
+          for (const mutation of batch.mutations) {
+            const key = mutation.key;
+            // Only process documents belonging to the collection.
+            if (!query.path.isImmediateParentOf(key.path)) {
+              continue;
+            }
+            if (
+              mutation instanceof PatchMutation &&
+              results.get(mutation.key) === null
+            ) {
+              missingBaseDocEntriesForPatching = missingBaseDocEntriesForPatching.add(
+                mutation.key
+              );
+            }
+          }
+        }
+        return this.remoteDocumentCache.getEntries(
+          transaction,
+          missingBaseDocEntriesForPatching
+        );
+      })
+      .next(baseDocsForPatching => {
```

Comment:
Nit: this variable name is perhaps slightly misleading; without context, it sounds like it contains _all_ base documents, but in fact it's only a subset of those documents (those that have pending writes making them potentially match). Consider renaming (`docsWithPendingWrites`? `docsWithLocalPatches`?).

Reply:
changed to docsWithPendingPatches.